### ISSUE TITLE
Documentation effort

### ISSUE THREAD

**Issue by Geal** [Geal/nom#522]
As we know, the [nom documentation](https://docs.rs/nom) is not that great. It has been improving, but not fast enough, and not with a satisfying quality. It's getting to the point where [anytime someone talks about nom, someone else chimes in to say the documentation is crap](https://www.reddit.com/r/rust/comments/6efmov/rust_toronto_meetup_introduction_to_parsing_using/). This is unacceptable and a big failure on my part.

So, let's improve things:
- the [first page](https://docs.rs/nom/3.0.0/nom/index.html) is daunting. There's a big example that uses advanced features like `fold_many0`, that does not help comprehension. Maybe the [JSON example](https://github.com/Geal/nom/blob/master/tests/json.rs) would be better? (provided we remove the float parsers and replace with the builtin ones). There should be an obvious path: "where do I go if I want to parse this kind of format?" "I know how to create a first parser, how do I find the combinators I need?" "I got this error, how do I fix it?"
- the big list of combinators is not helpful to beginners. Once you know how nom works, displaying a kind of function definition works to explain what it does, but it's not enough in a lot of cases. It does not help that we have limited room for the text of each macro on the first page. We could add more documentation inside modules, like the [whitespace module](https://docs.rs/nom/3.0.0/nom/whitespace/index.html), but how do we make sure beginners will check those out instead of going directly to the macros list?
- the [guides](http://rust.unhandledexpression.com/nom/home.html) have broken links. They're available on my own doc hosting, but docs.rs cannot render them, so the links are broken.
- the story around `named!`, `named_args!`, custom functions, methods, is confusing. Those combinators were designed to hide how the input is passed to parsers and reduce confusion, but they do not help much. Since they're the first combinators people will use, they need to be explained clearly.
- most combinators only have a one liner as documentation. I don't think large code examples will help so much there, they don't improve readability. But some small examples, with maybe some kind of table with the expected result for some inputs, would be better to get a feel of how combinators work.

This will be a lot of work, and I probably won't be able to do it all alone, but it has to be done.

**Comment by sharazam** [comment]
@Geal Like I said on reddit, it needs a lot of work. I also said that I'd do a PR, but for now I've just rewritten the lib.rs file - if you could look over it, that'd be nice: https://github.com/sharazam/nom-documentation-readme/blob/master/README.md

It's not finished, largely based on a transcript of the Toronto video.

I think it would be OK to make a shortcodes for parsers in the macros first sentence (to deal with the space issue). Like: "`call!(p1)` runs parser p1" or something like that - without the huge `IResult` thing.
 
For example, you told me to use the `u64!` macro, but I cannot for my life figure out how to use it, everything results in an error. The documentation says that it takes the endianness as an argument, but doesn't say if it's the first or second argument or how to use it. All I ended up doing was this: http://play.integer32.com/?gist=4ba80134e419b2cbbbe4191d404194ce&version=stable

The docs for the `u64!` macro say that it has two inputs: the bytes it operates on and the endianness. But it doesn't work and I don't understand how I can insert the endianness that I want.

Basically, all I want to know is what parsers are "compatible" and how I have to (syntactically) combine them to do what I want. I surely don't need hand-holding with everything, but the learning curve is very steep right now.

**Comment by ekse** [comment]
Regarding `u64!`, calling `u64!(endianness)` should work in your code. The input / output parameters are implicit when using a macro this way.

**Comment by sharazam** [comment]
@ekse Thanks for the help, but why is this not documented when what options are implicit. How should I have guessed that?

Any comments regarding my proposal for the main documentation? https://github.com/sharazam/nom-documentation-readme/blob/master/README.md

**Comment by jsgf** [comment]
This is great! I was just about to file an issue about this, and I'm really glad you're on top of this. I agree with all the points.

Some specific things that come to mind:

- Every combinator should have a summary line trying to describe it; skip the signature because they're not easily readable. The main point is to make it possible to scan the list and find something that matches the problem at hand.
- Every combinator should have a prose description of its function and a code example for each of its variants. The macro pattern soup rustdoc generates is useless. (This is mentioned in the OP, but I think its worth repeating, esp covering all the variants.)
- There need to be some larger worked examples which are not trivial "infix calculator" level, but touch on techniques like using custom error types, streaming, etc. Pointing to 3rd-party projects using nom is great, but they're not necessarily good teaching examples, using most appropriate techniques, or even up to date with the API. And even if they are good in all those ways, its hard to find examples that match something specific.

**Comment by Geal** [comment]
@sharazam this looks good, but I think I will actually reduce the readme to more general statements, like what you can achieve with nom, and move what looks more like a tutorial to the crate level docs. It's possible to make more long form documents in https://github.com/Geal/nom/tree/master/doc and they'll end up on http://rust.unhandledexpression.com/nom/home.html (docs.rs does not support that way of generating docs yet).
Putting the tutorial in crate docs makes it easier to keep in sync with the code. Readme examples are not compiled and tested.
A big issue right now is compiling a list of the combinators and parsers by context. Something like "use those for characters", "use those for parsing a sequence of things", etc.

**Comment by willi-kappler** [comment]
I've read through the macros and here are the ones that lack an example:

- [x] alt_complete
- [ ] apply_m
- [ ] call_m
- [ ] char
- [ ] closure
- [ ] cond_with_error
- [ ] consumer_from_parser
- [ ] eat_separator
- [ ] eof
- [ ] error_code
- [ ] error_node
- [ ] error_node_position
- [ ] error_position
- [ ] expr_res
- [ ] flat_map
- [ ] i16
- [ ] i32
- [ ] i64
- [ ] length_bytes
- [ ] length_count
- [ ] length_data
- [ ] length_value
- [ ] map
- [ ] map_opt
- [ ] map_res
- [ ] named_args
- [ ] none_of
- [ ] one_of
- [ ] pair
- [ ] parse_to
- [ ] preceded
- [ ] sep
- [ ] separated_list
- [ ] separated_list_complete
- [ ] separated_nonempty_list
- [ ] separated_nonempty_list_complete
- [ ] separated_pair
- [ ] tag_bits
- [ ] take_str
- [ ] take_till
- [ ] take_till1
- [ ] take_till1_s
- [ ] take_till_s
- [ ] take_until
- [ ] take_until1
- [ ] take_until_and_consume
- [ ] take_until_and_consume1
- [ ] take_until_and_consume_s
- [ ] take_until_either
- [ ] take_until_either_and_consume
- [ ] take_until_s
- [ ] take_while1
- [ ] take_while1_s -> broken
- [ ] terminated
- [ ] u16
- [ ] u32
- [ ] u64
- [ ] wrap_sep


I can start to work on these.
As Geal said for some a table with input / output would also be helpful.

**Comment by willi-kappler** [comment]
And here is a list of functions that lack an example (and sometimes even a documentation)

- [ ] alpha
- [ ] alphanumeric
- [ ] anychar
- [ ] be_f32
- [ ] be_f64
- [ ] be_i16
- [ ] be_i24
- [ ] be_i32
- [ ] be_i64
- [ ] be_i8
- [ ] be_u16
- [ ] be_u24
- [ ] be_u32
- [ ] be_u64
- [ ] be_u8
- [ ] begin
- [ ] code_from_offset
- [ ] crlf
- [ ] digit
- [ ] double
- [ ] double_s
- [ ] eol
- [ ] error_to_u32
- [ ] float
- [ ] float_s
- [ ] hex_digit
- [ ] hex_u32
- [ ] is_alphabetic
- [ ] is_alphanumeric
- [ ] is_digit
- [ ] is_hex_digit
- [ ] is_oct_digit
- [ ] is_space
- [ ] le_f32
- [ ] le_f64
- [ ] le_i16
- [ ] le_i24
- [ ] le_i32
- [ ] le_i64
- [ ] le_i8
- [ ] le_u16
- [ ] le_u24
- [ ] le_u32
- [ ] le_u64
- [ ] le_u8
- [ ] line_ending
- [ ] multispace
- [ ] newline
- [ ] non_empty
- [ ] not_line_ending
- [ ] oct_digit
- [ ] print
- [ ] print_codes
- [ ] reset_color
- [ ] rest
- [ ] rest_s
- [ ] shift
- [ ] sized_buffer
- [ ] slice_to_offsets
- [ ] space
- [ ] tab
- [ ] tag_cl
- [ ] write_color

**Comment by willi-kappler** [comment]
I've also noted that some examples use

    ```rust,ignore


and some use

    ```ignore

which one should be prefered ?

Ideally all examples should run as test cases. I can try to do this for the new ones and move the existing ones afterwards.

**Comment by bheisler** [comment]
I took a shot at improving the documentation starting with the bits module at https://github.com/bheisler/nom/blob/master/src/bits.rs.

So far I've:
* Added a short (prose) description to the beginning of the documentation for each macro, which shows up on the main index page. This should help with finding the macro that does what you need. It also moves the detailed signature to the macro-specific page along with any other detailed notes.
* Modified the examples to be more understandable
* Added an example for tag_bits, which didn't have one.

If this sounds good to the maintainer(s) and if time permits, I'll carry on to make the same changes for the other files and start submitting pull requests.

**Comment by Geal** [comment]
closing this in favor of #914
